Add support for features

MikaelSmith opened the pull request:
Adds support for volumefs and transport features.

MikaelSmith committed:
Add support for features

Adds support for volumefs and transport features.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith commented:
Example output with these new features
```
{
  "type_id": "Boltwash",
  "name": "bolt",
  "methods": [
    [
      "list",
      [
        {
          "type_id": "Group",
          "name": "all",
          "methods": [
            [
              "list",
              [
                {
                  "type_id": "Target",
                  "name": "pe-master.platform9.puppet.net",
                  "methods": [
                    [
                      "list",
                      [
                        {
                          "type_id": "__volume::fs__",
                          "name": "fs",
                          "state": "{\"maxdepth\":2}"
                        }
                      ]
                    ],
                    [
                      "exec",
                      {
                        "transport": "ssh",
                        "options": {
                          "host": "pe-master.platform9.puppet.net",
                          "user": "ubuntu"
                        }
                      }
                    ],
                    "schema"
                  ],
                  "partial_metadata": {
                    "name": "pe-master.platform9.puppet.net",
                    "config": {
                      "transport": "ssh",
                      "ssh": {
                        "connect-timeout": 10,
                        "tty": false,
                        "load-config": true,
                        "disconnect-timeout": 5,
                        "user": "ubuntu",
                        "run-as": "root",
                        "host-key-check": false
                      }
                    }
                  },
                  "state": "{\"connection_info\":{\"connect-timeout\":10,\"tty\":false,\"load-config\":true,\"disconnect-timeout\":5,\"user\":\"ubuntu\",\"run-as\":\"root\",\"host-key-check\":false,\"name\":\"pe-master.platform9.puppet.net\",\"uri\":\"pe-master.platform9.puppet.net\",\"protocol\":\"ssh\",\"password\":null,\"host\":\"pe-master.platform9.puppet.net\",\"port\":null},\"klass\":\"Target\",\"name\":\"pe-master.platform9.puppet.net\"}"
                },
                {
                  "type_id": "Target",
                  "name": "127.0.0.1:20022",
                  "methods": [
                    [
                      "list",
                      [
                        {
                          "type_id": "__volume::fs__",
                          "name": "fs",
                          "state": "{\"maxdepth\":2}"
                        }
                      ]
                    ],
                    [
                      "exec",
                      {
                        "transport": "ssh",
                        "options": {
                          "host": "127.0.0.1:20022",
                          "user": "bolt",
                          "password": "bolt"
                        }
                      }
                    ],
                    "schema"
                  ],
                  "partial_metadata": {
                    "name": "127.0.0.1:20022",
                    "config": {
                      "transport": "ssh",
                      "ssh": {
                        "connect-timeout": 10,
                        "tty": false,
                        "load-config": true,
                        "disconnect-timeout": 5,
                        "user": "bolt",
                        "password": "bolt",
                        "host-key-check": false,
                        "run-as": "root",
                        "sudo-password": "bolt"
                      }
                    }
                  },
                  "state": "{\"connection_info\":{\"connect-timeout\":10,\"tty\":false,\"load-config\":true,\"disconnect-timeout\":5,\"user\":\"bolt\",\"password\":\"bolt\",\"host-key-check\":false,\"run-as\":\"root\",\"sudo-password\":\"bolt\",\"name\":\"127.0.0.1:20022\",\"uri\":\"127.0.0.1:20022\",\"protocol\":\"ssh\",\"host\":\"127.0.0.1\",\"port\":20022},\"klass\":\"Target\",\"name\":\"127.0.0.1:20022\"}"
                },
                {
                  "type_id": "Target",
                  "name": "127.0.0.1:25986",
                  "methods": [
                    [
                      "list",
                      [
                        {
                          "type_id": "__volume::fs__",
                          "name": "fs",
                          "state": "{\"maxdepth\":2}"
                        }
                      ]
                    ],
                    "exec",
                    "schema"
                  ],
                  "partial_metadata": {
                    "name": "127.0.0.1:25986",
                    "config": {
                      "transport": "winrm",
                      "winrm": {
                        "connect-timeout": 10,
                        "ssl": true,
                        "ssl-verify": false,
                        "file-protocol": "winrm",
                        "user": "bolt",
                        "password": "bolt"
                      }
                    }
                  },
                  "state": "{\"connection_info\":{\"connect-timeout\":10,\"ssl\":true,\"ssl-verify\":false,\"file-protocol\":\"winrm\",\"user\":\"bolt\",\"password\":\"bolt\",\"name\":\"127.0.0.1:25986\",\"uri\":\"127.0.0.1:25986\",\"protocol\":\"winrm\",\"host\":\"127.0.0.1\",\"port\":25986},\"klass\":\"Target\",\"name\":\"127.0.0.1:25986\"}"
                }
              ]
            ],
            "schema"
          ],
          "state": "{\"klass\":\"Group\",\"name\":\"all\"}"
        },
        {
          "type_id": "Group",
          "name": "linux",
          "methods": [
            [
              "list",
              [
                {
                  "type_id": "Target",
                  "name": "pe-master.platform9.puppet.net",
                  "methods": [
                    [
                      "list",
                      [
                        {
                          "type_id": "__volume::fs__",
                          "name": "fs",
                          "state": "{\"maxdepth\":2}"
                        }
                      ]
                    ],
                    [
                      "exec",
                      {
                        "transport": "ssh",
                        "options": {
                          "host": "pe-master.platform9.puppet.net",
                          "user": "ubuntu"
                        }
                      }
                    ],
                    "schema"
                  ],
                  "partial_metadata": {
                    "name": "pe-master.platform9.puppet.net",
                    "config": {
                      "transport": "ssh",
                      "ssh": {
                        "connect-timeout": 10,
                        "tty": false,
                        "load-config": true,
                        "disconnect-timeout": 5,
                        "user": "ubuntu",
                        "run-as": "root",
                        "host-key-check": false
                      }
                    }
                  },
                  "state": "{\"connection_info\":{\"connect-timeout\":10,\"tty\":false,\"load-config\":true,\"disconnect-timeout\":5,\"user\":\"ubuntu\",\"run-as\":\"root\",\"host-key-check\":false,\"name\":\"pe-master.platform9.puppet.net\",\"uri\":\"pe-master.platform9.puppet.net\",\"protocol\":\"ssh\",\"password\":null,\"host\":\"pe-master.platform9.puppet.net\",\"port\":null},\"klass\":\"Target\",\"name\":\"pe-master.platform9.puppet.net\"}"
                }
              ]
            ],
            "schema"
          ],
          "state": "{\"klass\":\"Group\",\"name\":\"linux\"}"
        },
        {
          "type_id": "Group",
          "name": "docker",
          "methods": [
            [
              "list",
              [
                {
                  "type_id": "Target",
                  "name": "127.0.0.1:20022",
                  "methods": [
                    [
                      "list",
                      [
                        {
                          "type_id": "__volume::fs__",
                          "name": "fs",
                          "state": "{\"maxdepth\":2}"
                        }
                      ]
                    ],
                    [
                      "exec",
                      {
                        "transport": "ssh",
                        "options": {
                          "host": "127.0.0.1:20022",
                          "user": "bolt",
                          "password": "bolt"
                        }
                      }
                    ],
                    "schema"
                  ],
                  "partial_metadata": {
                    "name": "127.0.0.1:20022",
                    "config": {
                      "transport": "ssh",
                      "ssh": {
                        "connect-timeout": 10,
                        "tty": false,
                        "load-config": true,
                        "disconnect-timeout": 5,
                        "user": "bolt",
                        "password": "bolt",
                        "host-key-check": false,
                        "run-as": "root",
                        "sudo-password": "bolt"
                      }
                    }
                  },
                  "state": "{\"connection_info\":{\"connect-timeout\":10,\"tty\":false,\"load-config\":true,\"disconnect-timeout\":5,\"user\":\"bolt\",\"password\":\"bolt\",\"host-key-check\":false,\"run-as\":\"root\",\"sudo-password\":\"bolt\",\"name\":\"127.0.0.1:20022\",\"uri\":\"127.0.0.1:20022\",\"protocol\":\"ssh\",\"host\":\"127.0.0.1\",\"port\":20022},\"klass\":\"Target\",\"name\":\"127.0.0.1:20022\"}"
                }
              ]
            ],
            "schema"
          ],
          "state": "{\"klass\":\"Group\",\"name\":\"docker\"}"
        },
        {
          "type_id": "Group",
          "name": "nano",
          "methods": [
            [
              "list",
              [
                {
                  "type_id": "Target",
                  "name": "127.0.0.1:25986",
                  "methods": [
                    [
                      "list",
                      [
                        {
                          "type_id": "__volume::fs__",
                          "name": "fs",
                          "state": "{\"maxdepth\":2}"
                        }
                      ]
                    ],
                    "exec",
                    "schema"
                  ],
                  "partial_metadata": {
                    "name": "127.0.0.1:25986",
                    "config": {
                      "transport": "winrm",
                      "winrm": {
                        "connect-timeout": 10,
                        "ssl": true,
                        "ssl-verify": false,
                        "file-protocol": "winrm",
                        "user": "bolt",
                        "password": "bolt"
                      }
                    }
                  },
                  "state": "{\"connection_info\":{\"connect-timeout\":10,\"ssl\":true,\"ssl-verify\":false,\"file-protocol\":\"winrm\",\"user\":\"bolt\",\"password\":\"bolt\",\"name\":\"127.0.0.1:25986\",\"uri\":\"127.0.0.1:25986\",\"protocol\":\"winrm\",\"host\":\"127.0.0.1\",\"port\":25986},\"klass\":\"Target\",\"name\":\"127.0.0.1:25986\"}"
                }
              ]
            ],
            "schema"
          ],
          "state": "{\"klass\":\"Group\",\"name\":\"nano\"}"
        }
      ]
    ],
    [
      "schema",
      {
        "Boltwash": {
          "label": "bolt",
          "methods": [
            "list",
            "schema"
          ],
          "singleton": true,
          "children": [
            "Group"
          ]
        },
        "Group": {
          "label": "group",
          "methods": [
            "list",
            "schema"
          ],
          "children": [
            "Target"
          ]
        },
        "Target": {
          "label": "target",
          "methods": [
            "list",
            "exec",
            "schema"
          ],
          "children": [
            "__volume::fs__"
          ]
        }
      }
    ]
  ],
  "state": "{\"klass\":\"Boltwash\",\"name\":\"bolt\"}"
}
```

ekinanp commented on lib/wash/entry.rb:
```diff
@@ -343,6 +354,50 @@ def cache_ttls(ttls = {})
+    # @param [Symbol] klass The transport to use. Only `:ssh` is supported
+    # @param [Hash] options For SSH: :host (required), :user, :password,
+    #     :fallbackuser, :identityfile, :knownhosts, :hostkeyalias, :retries
+    def transport(klass, options = {})
```
We could define the `exec` for them and save some clutter.

MikaelSmith replied:
So that's challenging because the `exec` is part of the class and can be used in defining the schema, but the `transport` calls happen during instance initialization. I'd still have to have something called during class definition.

ekinanp replied:
Ahh yeah you're right, we do need a class-level defn. for entry schemas so your example's correct. Would be nice to add a comment about that somewhere because it seems like that'd be a common question with a subtle answer.

MikaelSmith replied:
That's right in the code comment about 12 lines up.

ekinanp commented on lib/wash/entry.rb:
```diff
@@ -343,6 +354,50 @@ def cache_ttls(ttls = {})
     end
 
+    # volumefs returns an entry that provides a view of the parent entry's filesystem.
+    # Use it when listing children. It can be used with `prefetch :list`.
```
I'd probably just say `volumefs` creates the `volume::fs` entry as specified in the external plugin docs. The current wording here makes it seem like a `wash-ruby` thing when it is more general than that.

Also, is mentioning `prefetch :list` necessary? Maybe a nit on my end, but when I read that, my brain forgets that this is just a regular entry JSON. I think your example is clear enough and shows that you can use `prefetch :list` with it. It might be nice to add one more entry in the `list` call just so "volumefs is a regular entry" is really codified

ekinanp reviewed (commented):
Looks good overall, just a few comments.

MikaelSmith committed:
Update rubygems api token

Hopefully fixes #18.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Clarify README and fix link to entry schemas

Signed-off-by: Michael Smith <michael.smith@puppet.com>

MikaelSmith committed:
Add spec tests

Adds spec tests for the `wash` library. Also update CI config to have
separate spec test and wash validate jobs, as well as separate out test
and gem release stages.

Signed-off-by: Michael Smith <michael.smith@puppet.com>

ekinanp commented on lib/wash/streamer.rb:
```diff
@@ -17,8 +17,8 @@ def write(chunk)
         end
-        STDOUT.print(chunk)
-        STDOUT.flush
+        $stdout.print(chunk)
```
Why this change?

MikaelSmith replied:
> Note: to_stdout and to_stderr work by temporarily replacing $stdout or $stderr, so
they're not able to intercept stream output that explicitly uses STDOUT/STDERR or that
uses a reference to $stdout/$stderr that was stored before the matcher is used.

From https://relishapp.com/rspec/rspec-expectations/docs/built-in-matchers/output-matcher.

ekinanp replied:
Ah ok, makes sense. Thanks.

ekinanp commented on spec/wash_spec.rb:
```diff
@@ -0,0 +1,27 @@
+require 'wash'
+
+describe Wash do
+  describe 'pretty_print?' do
```
Yeah, I was an idiot for making individual config fields helpers. It'd be nice to (at some point) refactor this into a hash that can be passed into a `Wash.config` type helper, but that's not that important right now.

ekinanp commented on README.md:
````diff
@@ -75,6 +75,34 @@ implements `stream` and `exec`. The calling conventions and return parameters fo
+end
+```
+
+_Exec implementations_ provide an implementation of the `exec` method so you don't need to define your own. Wash provides an _exec implementation_ of the SSH transport. See the [external plugin exec docs](https://puppetlabs.github.io/wash/docs/external-plugins#exec) for more on _exec implementations_, including all the options available for the SSH transport. Request this transport with:
````
Maybe a nit, but I'd (slightly) reword this to something like

> _Exec implementations_ provide an implementation of the `exec` method so you don't need to define your own. For example if your entry works with SSH, then you can use Wash's SSH transport to implement its `exec` method. \<everything else remains the same\>

Reason: "Wash provides an _exec implementation_ of the SSH transport" feels a bit abstract to me, unless it's easy for people to figure out that that sentence means "Hey, if my thing works with SSH I can use Wash's SSH stuff to implement my exec for me"

ekinanp commented:
I think the last commit may need to be re-worded -- it still mentions that features need to be included in the entry schema

MikaelSmith commented:
I'm working on some more testing for this PR. Not very simple to add integration tests for the new Wash features, so I'm focusing on spec tests.

MikaelSmith commented:
Finished.
